src/Build/BackEnd/Components/Logging/LoggingService.cs

```diff
@@ -900,8 +900,28 @@ public void PacketReceived(int node, INodePacket packet)
             }
 
             LogMessagePacket loggingPacket = (LogMessagePacket)packet;
-            InjectNonSerializedData(loggingPacket);
-            ProcessLoggingEvent(loggingPacket.NodeBuildEvent);
+
+            if (loggingPacket.EventType == LoggingEventType.CustomEvent
+                && ChangeWaves.AreFeaturesEnabled(ChangeWaves.Wave17_8)
+                && Traits.Instance.EscapeHatches.EnableWarningOnCustomBuildEvent)
+            {
+                BuildEventArgs buildEvent = loggingPacket.NodeBuildEvent.Value.Value;
+
+                // Serializing unknown CustomEvent which has to use unsecure BinaryFormatter by TranslateDotNet<T>
+                // Since BinaryFormatter is going to be deprecated, log warning so users can use new Extended*EventArgs instead of custom
+                // EventArgs derived from existing EventArgs.
+                LogWarning(
+                    buildEvent?.BuildEventContext ?? BuildEventContext.Invalid,
+                    null,
+                    BuildEventFileInfo.Empty,
+                    "DeprecatedEventSerialization",
+                    buildEvent?.GetType().Name ?? string.Empty);
+            }
+            else
+            {
+                InjectNonSerializedData(loggingPacket);
+                ProcessLoggingEvent(loggingPacket.NodeBuildEvent);
+            }
```

Comment:
Should we rather do this check before the serialization attempt? As otherwise if `EnableUnsafeBinaryFormatterSerialization` is not opted in, the build can crash on the serialization attemt and the warning (that should be a friendlier alternative) would not even get emitted. @rokonec - thoughts?

Plus if the custom events are discarded - the warning message should explicitly mention that (as it's behavioral change).

Reply:
The current adjustment looks good 👍 https://github.com/dotnet/msbuild/commit/68c52d550ae21ced86ba677962effc4cb3cc8bc3